toplevel/topcommon.ml

```diff
@@ -0,0 +1,302 @@
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Definitions for the interactive toplevel loop that are common between
+   bytecode and native *)
+
+open Format
+open Parsetree
+open Outcometree
+open Ast_helper
+
+(* Hooks for parsing functions *)
+
+let parse_toplevel_phrase = ref Parse.toplevel_phrase
+let parse_use_file = ref Parse.use_file
+let print_location = Location.print_loc
+let print_error = Location.print_report
+let print_warning = Location.print_warning
+let input_name = Location.input_name
+
+let parse_mod_use_file name lb =
+  let modname =
+    String.capitalize_ascii
+      (Filename.remove_extension (Filename.basename name))
+  in
+  let items =
+    List.concat
+      (List.map
+         (function Ptop_def s -> s | Ptop_dir _ -> [])
+         (!parse_use_file lb))
+  in
+  [ Ptop_def
+      [ Str.module_
+          (Mb.mk
+             (Location.mknoloc (Some modname))
+             (Mod.structure items)
+          )
+       ]
+   ]
+
+(* Hooks for printing *)
+
+let max_printer_depth = ref 100
+let max_printer_steps = ref 300
+
+let print_out_value = Oprint.out_value
+let print_out_type = Oprint.out_type
+let print_out_class_type = Oprint.out_class_type
+let print_out_module_type = Oprint.out_module_type
+let print_out_type_extension = Oprint.out_type_extension
+let print_out_sig_item = Oprint.out_sig_item
+let print_out_signature = Oprint.out_signature
+let print_out_phrase = Oprint.out_phrase
+
+
+(* The current typing environment for the toplevel *)
+
+let toplevel_env = ref Env.empty
+
+let backtrace = ref None
+
+(* Generic printer *)
+
+module type PRINTER = sig
+
+  module Printer: Genprintval.S
+
+  val print_value: Env.t -> Printer.t -> formatter -> Types.type_expr -> unit
+  val print_untyped_exception: formatter -> Printer.t -> unit
+
+  val print_exception_outcome : formatter -> exn -> unit
+
+  val outval_of_value:
+    Env.t -> Printer.t -> Types.type_expr -> Outcometree.out_value
+
+  type ('a, 'b) gen_printer =
+    | Zero of 'b
+    | Succ of ('a -> ('a, 'b) gen_printer)
+
+  val install_printer :
+    Path.t -> Types.type_expr -> (formatter -> Printer.t -> unit) -> unit
+  val install_generic_printer :
+    Path.t -> Path.t ->
+    (int -> (int -> Printer.t -> Outcometree.out_value,
+            Printer.t-> Outcometree.out_value) gen_printer) -> unit
+  val install_generic_printer' :
+    Path.t -> Path.t -> (formatter -> Printer.t -> unit,
+                         formatter -> Printer.t -> unit) gen_printer -> unit
+  val remove_printer : Path.t -> unit
+
+end
+
+module MakePrinter
+    (O: Genprintval.OBJ)
+    (Printer: Genprintval.S with type t = O.t)
+  : PRINTER with type Printer.t = O.t
```

Comment:
why the `O` parameter here ? It's filled by the `Obj` module in both cases ...

Reply:
Indeed, I just forwarded the argument to `Genprintval.Make`, but that could be simplified here.